bluetooth: tester: audio: Add verification for bis stopping

Frodevan opened the pull request:
Adds a semaphor to make sure the stream has properly been stopped by the
controller before returning from btp_bap_broadcast_source_stop.

fixes: https://github.com/zephyrproject-rtos/zephyr/issues/73411

Thalley commented on tests/bluetooth/tester/src/audio/btp/btp_bap.h:
```diff
@@ -274,3 +274,11 @@ struct btp_bap_pa_sync_req_ev {
+	bt_addr_le_t address;
+	uint8_t broadcast_id[BT_AUDIO_BROADCAST_ID_SIZE];
+	uint8_t bis_id;
+	uint8_t bis_stop_reason;
```
```suggestion
	uint8_t reason;
```

Minor suggestion: `bis_stop` is kind of implied by the name of this struct :)

Frodevan replied:
Allright, this is now changed.

Thalley commented on tests/bluetooth/tester/src/audio/btp_bap_broadcast.c:
```diff
@@ -125,15 +125,30 @@ static void stream_started(struct bt_bap_stream *stream)
 	btp_bap_audio_stream_stopped(&b_stream->audio_stream);
-
 	b_stream->bis_synced = false;
+	broadcaster = &remote_broadcast_sources[b_stream->source_id];
```
I think it's missing the variable declaration for `broadcaster`: 
```c
struct btp_bap_broadcast_remote_source *broadcaster;
```

Frodevan replied:
Missed that one, thanks

Thalley commented on tests/bluetooth/tester/src/audio/btp_bap_broadcast.c:
```diff
@@ -125,15 +125,30 @@ static void stream_started(struct bt_bap_stream *stream)
 	btp_send_bis_syced_ev(&broadcaster->address, broadcaster->broadcast_id, b_stream->bis_id);
 }
 
+static void btp_send_bis_stopped_ev(const bt_addr_le_t *address, uint32_t broadcast_id,
+				  uint8_t bis_id, uint8_t reason)
```
```suggestion
static void btp_send_bis_stopped_ev(const bt_addr_le_t *address, uint32_t broadcast_id,
				    uint8_t bis_id, uint8_t reason)
```

Frodevan replied:
Right, my editor uses a different standard indentation, but updated to look nice on git now.

Thalley replied:
If your editor supports https://editorconfig.org/ then it's easy to ensure that it works with many projects out of the box :)

Frodevan committed:
bluetooth: tester: audio: Add verification for bis stopping

Adds a semaphor to make sure the stream has properly been stopped by the
controller before returning from btp_bap_broadcast_source_stop.

Signed-off-by: Frode van der Meeren <frode.vandermeeren@nordicsemi.no>

Thalley commented on tests/bluetooth/tester/src/audio/btp_bap_broadcast.c:
```diff
@@ -484,13 +487,22 @@ uint8_t btp_bap_broadcast_source_stop(const void *cmd, uint16_t cmd_len,
 		return BTP_STATUS_FAILED;
 	}
 
+	err = k_sem_take(&sem_bis_stopped, K_MSEC(2000));
+	if (err != 0) {
+		LOG_DBG("Timed out waiting to stop BIS stream");
+
+		return BTP_STATUS_FAILED;
+	}
```
`stream_stopped` that gives `sem_bis_stopped` may be called multiple times. 

You should take the same number of semaphores as we have streams.

Frodevan replied:
This should now be resolved, given bt_bap_broadcast_source_stop stops broadcast source streams only.

sjanc commented:
#AutoPTS run zephyr BAP/BSRC/SCC/BV-34-C

sjanc commented on tests/bluetooth/tester/src/audio/btp_bap_broadcast.c:
```diff
@@ -484,13 +487,24 @@ uint8_t btp_bap_broadcast_source_stop(const void *cmd, uint16_t cmd_len,
 		return BTP_STATUS_FAILED;
 	}
 
+	for (int i = 0; i < ARRAY_SIZE(source->streams); i++) {
```
this should check for actual number of streams, not max supported, shouldn't it?

Frodevan replied:
Right, must this be done with stream->in_use, or is this number stored in the struct?

Thalley replied:
> Right, must this be done by checking stream->in_use, or is this number stored somewhere in the source struct?

I don't think we currently store it anywhere, but it's pretty easy to just add a counter in the `struct btp_bap_broadcast_remote_source` so that we always know how many `sem_stream_stopped` we should expect if we ever start fewer than `ARRAY_SIZE(source->streams)` streams

@sjanc any input?

Thalley replied:
`stream_stopped` will be called per stream, regardless of the number of subgroups

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/79329#issuecomment-2391187947, board: nrf53, estimated start time: 13:34:00, test case count: 1, estimated duration: 0:00:20<details><summary>Test cases to be run</summary>BAP/BSRC/SCC/BV-34-C</details>

codecoup-tester commented:
AutoPTS Bot results:
No failed test found.
<details><summary>Successful tests (1)</summary>BAP     BAP/BSRC/SCC/BV-34-C            PASS</details>

Thalley reviewed (dismissed):
Looks OK to me. Not sure about the timeout of `2000` ms, but I'll leave that to @sjanc :)

Frodevan commented:
My reasoning for the timeout is that I have not checked if there is a timeout for this call in Auto-PTS, so an infinite wait here has potentially far greater consequences for a (very long) test run than just an error-code returned during a test-case.

sjanc commented:
AutoPTS has 2s timeout for BTP response (after that it assumes IUT is dead), so I suggest having 1s timeout here (it should be more than enough anyway)

Frodevan commented:
Timeout adjusted

sjanc commented on tests/bluetooth/tester/src/audio/btp_bap_broadcast.c:
```diff
@@ -484,13 +489,24 @@ uint8_t btp_bap_broadcast_source_stop(const void *cmd, uint16_t cmd_len,
 	}
 
+	for (int i = 0; i < source->stream_count; i++) {
+		err = k_sem_take(&sem_stream_stopped, K_MSEC(1000));
```
to be very strict that timeout should be for all semaphores,  but since this will indicate bug in stack I don't think it matter much and adding time tracking would be overkill :-)
